contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol

```diff
@@ -0,0 +1,388 @@
+        for (uint256 i = 0; i < tokens.length; i++) {
+            IERC20(tokens[i]).safeApprove(address(uniswapSwapRouter), 0);
+            IERC20(tokens[i]).safeApprove(
+                address(uniswapSwapRouter),
+                uint256(-1)
+            );
+        }
+    }
+
+    function _swapWithUni(
+        address sourceTokenAddress,
+        address destTokenAddress,
+        address receiverAddress,
+        uint256 minSourceTokenAmount,
+        uint256 maxSourceTokenAmount,
+        uint256 requiredDestTokenAmount,
+        bytes memory payload
+    )
+        internal
+        returns (uint256 sourceTokenAmountUsed, uint256 destTokenAmountReceived)
+    {
+        if (requiredDestTokenAmount != 0) {
+            IUniswapV3SwapRouter.ExactOutputParams[] memory exactParams = abi
+                .decode(payload, (IUniswapV3SwapRouter.ExactOutputParams[]));
+            bytes[] memory encodedTXs = new bytes[](exactParams.length);
+            uint256 totalAmountsOut = 0;
+            uint256 totalAmountsInMax = 0;
+            for (
+                uint256 uniqueOutputParam = 0;
+                uniqueOutputParam < exactParams.length;
+                uniqueOutputParam++
+            ) {
+                exactParams[uniqueOutputParam].recipient = receiverAddress; //sets receiver to this protocol
+                _verifyInputData(
+                    exactParams[uniqueOutputParam].path,
+                    sourceTokenAddress,
+                    destTokenAddress
+                );
+                totalAmountsOut = totalAmountsOut.add(
+                    exactParams[uniqueOutputParam].amountOut
+                );
+                totalAmountsInMax = totalAmountsInMax.add(
+                    exactParams[uniqueOutputParam].amountInMaximum
+                );
+                if (uniqueOutputParam == 0) {
+                    continue; //encodes the tx later
+                }
+                encodedTXs[uniqueOutputParam] = abi.encodeWithSelector(
+                    uniswapSwapRouter.exactOutput.selector,
+                    exactParams[uniqueOutputParam]
+                );
+            }
+            if (totalAmountsOut < requiredDestTokenAmount) {
+                //does not need safe math as it cannot overflow
+                uint256 displace = _numberAdjustment(
+                    totalAmountsOut,
+                    requiredDestTokenAmount
+                );
+                exactParams[0].amountOut += displace; //adds displacement to first swap set
+                totalAmountsOut += displace;
+            }
+            if (totalAmountsInMax < maxSourceTokenAmount) {
+                //does not need safe math as it cannot overflow
+                uint256 displace = _numberAdjustment(
+                    totalAmountsOut,
+                    maxSourceTokenAmount
+                );
+                exactParams[0].amountInMaximum += displace; //adds displacement to first swap set
```

Comment:
how do you know this won't overflow?

Reply:
displace is the difference between maxSourceTokenAmount and totalAmountsInMax and exactParams[0].amountInMaximum < totalAmountsInMax so it should not overflow at any point. I realize now that I used the wrong variable for _numberAdjustment call though.